test/middleware.js

```diff
@@ -49,6 +49,151 @@ describe('middleware', () => {
+          argv.mw.should.equal('mw')
+          return done()
+        }
+      )
+      .exitProcess(false) // defaults to true.
+      .parse()
+  })
+
+  it('runs the pre-check middleware and ensures theres a context object with commands and availableOptions', function (done) {
+    yargs(['mw'])
+      .middleware(function (argv, context) {
+        argv.mw = context.commands[0]
+        argv.other = Array.isArray(context.availableOptions.mw)
+      }, true)
+      .command(
+        'mw',
+        'adds func to middleware',
+        {
+          'mw': {
+            'demand': true,
+            'string': true
+          }
+        },
+        function (argv) {
+          // we should get the argv filled with data from the middleware
+          argv.mw.should.equal('mw')
+          argv.other.should.equal(true)
+          return done()
+        }
+      )
+      .exitProcess(false) // defaults to true.
+      .parse()
+  })
+
+  it('runs the pre-check middlware with an array passed in and ensures theres a context object with commands and availableOptions', function (done) {
+    yargs(['mw'])
+      .middleware([function (argv, context) {
+        argv.mw = context.commands[0]
+        argv.other = Array.isArray(context.availableOptions.mw)
+      }], true)
+      .command(
+        'mw',
+        'adds func to middleware',
+        {
+          'mw': {
+            'demand': true,
+            'string': true
+          }
+        },
+        function (argv) {
+          // we should get the argv filled with data from the middleware
+          argv.mw.should.equal('mw')
+          argv.other.should.equal(true)
+          return done()
+        }
+      )
+      .exitProcess(false) // defaults to true.
+      .parse()
+  })
+
+  it('runs the pre-check middlware ensures if an async function is ran it throws an error', function (done) {
+    try {
+      yargs(['mw'])
+        .middleware([async function (argv, context) {
+          argv.mw = context.commands[0]
+          argv.other = Array.isArray(context.availableOptions.mw)
+        }], true)
+        .command(
+          'mw',
+          'adds func to middleware',
+          {
+            'mw': {
+              'demand': true,
+              'string': true
+            }
+          },
+          function (argv) {
+            // we should get the argv filled with data from the middleware
+            argv.mw.should.equal('mw')
+            argv.other.should.equal(true)
+            return done(new Error('This should not have reached this point.'))
+          }
+        )
+        .exitProcess(false) // defaults to true.
+        .parse()
+    } catch (err) {
+      expect(err.message).to.equal('The passed in middleware with applyBeforeValidation set to true may not be used with async functions.')
+      done()
+    }
+  })
+
+  it('Ensure precheck middleware does not run non-precheck middleware, and vice versa', function (done) {
```

Comment:
I'd probably update the terminology in tests to use the term `applyBeforeValidation` rather than `pre-check;

`it should run middleware with applyBeforeValidation set to true before validation`

`it should run middleware with applyBeforeValidation set to false after validation`

I'm happy to do some tweaking like this myself before landing.

Reply:
I renamed things to before validation from pre-check